Optimize performance by optionally collectiong only EA directives

drpicox opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)** Performance optimization that introduces a new feature as `$compileProvider` configuration.

**What is the current behavior? (You can also link to an open issue here)**: `$compile` collect directives from entities, attributes, classes and comments. It parses all four kind of directives, although usually only E (components) and A ("directives") are used in most of the projects.

**What is the new behavior (if this is a feature change)?**: It introduces a new configuration in the `$compileProvider` that disables the compilation of C and M directives:

``` javascript
$compileProvider.commentDirectivesEnabled(false);
$compileProvider.cssClassDirectivesEnabled(false);
```

**Does this PR introduce a breaking change?**
No. By default is configured true, thus it is compatible with all existing code.

**Please check if the PR fulfills these requirements**
- [X] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

**Other information**:
It is closely related to https://github.com/angular/angular.js/pull/14848 
I have reused the same benchmarks to measure the improved performance of this PR, and it is around **10-15% of performance improvement** by setting restrictDirectivesTo to 'EA'.

drpicox committed:
perf($compile): add flags to compile css class and comment directives conditionally

petebacondarwin commented:
This seems like a pretty good idea since just about every Angular app I have ever written doesn't actually use comment or CSS class based directives. We could even make this opt-out in a future version...

The API naming seems a bit clunky though... Perhaps something more like `restrictDirectivesTo('EA')` or similar? Open to discussion on that...

Narretz commented:
I also like it. One of these things that are really obvious in hindsight.
I also think that we could expand the API to allow any combination of restricts (although EA is needed by probably 99.999% of apps).

drpicox commented:
Good!
I'll try to update it later today :)
(I also like more the naming)
On Thu, 30 Jun 2016 at 15:52, Martin Staffa <notifications@github.com>
wrote:

> I also like it. One of these things that are really obvious in hindsight.
> I also think that we could expand the API to allow any combination of
> restricts (although EA is needed by probably 99.999% of apps).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/14850#issuecomment-229663906>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACoWyR5We75osb8DzSoNncl7r3igCo_Lks5qQ8ozgaJpZM4JB9-c>
> .
>
-- 
Dr. David Ródenas
*Vice-degà *
Col·legi d’Enginyeria Informàtica de Catalunya

drpicox commented:
I have updated it.

The function name is `restrictDirectivesTo`, default value is `'EACM'`, and the implementation is almost the same.

Because:
- `restrictDirectivesTo` may accept any combination of EACM, 
- I do not want to add more conditionals than previously existing (deactivate E or A is a little bit more complex),
- and it makes no sense to do not include `EA`
I force to have at least `EA` in the expression.

petebacondarwin commented:
Great and rapid work - thanks @drpicox 
Sorry to be a pain but now that I realise that disabling E and A is not really viable, my suggested API naming doesn't make that much sense.

Can we have a quick chat about what might work best before you make any more changes...

Perhaps we could make it explicit, following the getter/setter style of `debugInfoEnabled`:

```
commentDirectivesEnabled(false);
cssClassDirectivesEnabled(false);
```

or perhaps we could disable them together...

```
commentAndCssClassDirectivesEnabled(false);
```

or some better thing??

drpicox commented:
I have seen some people using class directives, mainly because some people work like times of backbone or jquey and they use templates from designers which do thinks like <button class="save-user-button"></button>. There is also a popular library to draw charts by using canvas that has class directives (it break when restrict was defaulted to EA but they corrected it to support C). 

So it make sense to have separated class and comments. I have never seen a comment directive. I know why it is supposed to be useful because of the angular documentation. 

I have difficult to do a chat by now, now I'm at AngularCamp listening to @toddmotto talking about .component :)

drpicox commented:
Btw I'm follows you at Twitter, so if you follow me you can send me pms and have the chat.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -3128,17 +3128,129 @@ describe('$compile', function() {
       });
     });
 
-    it('should observe interpolated attrs', inject(function($rootScope, $compile) {
```
Why was this test removed?

drpicox replied:
Ops! It shouldn't!

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -3128,17 +3128,129 @@ describe('$compile', function() {
+        }));
+
+        var $compile, $rootScope;
+        beforeEach(inject(function(_$compile_,_$rootScope_) {
```
Since this blocks are repeated between all "child" `describe`s, you could move them to the "parent" `describe`.

Narretz commented:
I think the explicit version is the way to go. It's verbose, but you are not gonna use this more than once per app, and it also doesn't give the impression that you can disable EA
```
commentDirectivesEnabled(false);
cssClassDirectivesEnabled(false);
```

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -3128,17 +3128,129 @@ describe('$compile', function() {
+          $rootScope = _$rootScope_;
+        }));
+
+        it('should not compile class directives', function() {
```
I think it would be cleaner to use `test-collect` for all testcases. It makes the comparison easier.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -3128,17 +3128,129 @@ describe('$compile', function() {
+          expect(element.text()).toBe('text interpolated');
+        });
+
+        it('should interpolate expressions inside class attribute', function() {
```
:+1:

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -3128,17 +3128,129 @@ describe('$compile', function() {
+          }).toThrowMinErr('$compile', 'rstrto');
+        }));
+
+        afterEach(inject(function($compile) {
```
Why not `beforeEach`?

gkalpak commented:
Yeah, :+1: for the explicit versions from me too.

BTW, I would feel more confident about the tests, if we used the `they` helper to run each of them against all possible combinations and verfiy that the behavior is correct. Something like this:

```js
they('should handle comment directives appropriately ($prop)', [
  {restrict: 'EA', collect: false},
  {restrict: 'EAC', collect: false},
  {restrict: 'EAM', collect: true},
  {restrict: 'EACM', collect: true}
], function(config) {
  module(function($compileProvider) {
    $compileProvider.restrictDirectivesTo(config.restrict);
  });

  inject(function() {
    var html = '<!-- directive: test-collect -->';
    element = $compile('<div>' + html + '</div>')($rootScope);
    expect(collected).toBe(config.collect);
  });
});
```

petebacondarwin commented:
OK, so lets go with

```
commentDirectivesEnabled(false);
cssClassDirectivesEnabled(false);
```

@drpicox can you take a look at @gkalpak's comments about the testing?

Thanks

drpicox commented:
Ok. I'm working on now.

drpicox commented:
I like the idea of using `they` to explore all possibilities, but, what about instead of writing 5 theys use forEach with a describe?

drpicox commented:
About 1.6.x, if this flags change its default value, probably many of the tests will be affected: for example, all tests for class and comment directives will fail.

drpicox commented:
I'm just "fixing" the tests to disable css and comment directives by default.
I think that it is a good idea just to cach any possible bug or undeterminated behaviour that may happen. I go case by case enabling css or comment only if they are required by the test (but I'm not changing the tests to not test class directives if it is not the main objective of the test).

Narretz commented:
@drpicox do you have plans to update this PR again?

drpicox commented:
Hi, I think that's ok.
In fact in #14859 I did lots of testing enough to be sure that it would work well when only EA directives are activated.

So, it's ok!

Narretz commented on src/ng/compile.js:
```diff
@@ -1386,6 +1386,61 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+  var ignoreCDirectives = false;
+  /**
+   * @ngdoc method
+   * @name $compileProvider#commentDirectivesEnabled
```
should be `...#cssDirectivesEnabled`

Narretz commented on src/ng/compile.js:
```diff
@@ -1386,6 +1386,61 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+   *
+   * @param {boolean} false if the compiler may ignore comment directives
+   */
+  this.cssDirectivesEnabled = function(value) {
```
Actually, I think we decided this should be `cssClassDirectivesEnabled`

Narretz commented:
Hm, yes the testing was extensive, but we are gonna merge this PR, not the other one. Three things are still needed here:
1. Rebase on current master because of merge conflicts
2. Change the method name from `cssDirectivesEnabled` to `cssClassDirectivesEnabled`
2. A test that shows that comment and class directives directives are collected by default (i.e. with an untouched config)

Narretz commented:
@drpicox I can also make these changes, so just tell me if you'll do it or I should. :)

drpicox commented:
Thanks! 
I have already done the changes.
I'm passing all tests locally, I push them as soon as they are ok.

Narretz commented:
@drpicox Looking forward to the update!

drpicox commented:
Yep!
Next time I'll ping!

Thanks @Narretz !

Narretz commented on src/ng/compile.js:
```diff
@@ -1386,6 +1386,61 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
     return TTL;
   };
 
+  var ignoreMDirectives = false;
```
I'd like this to be explicit

drpicox replied:
what about {{disableCommentDirectives}} or {{commentDirectivesDisabled}}?
or do you prefer {{commentDirectivesEnabled}} ?

Narretz replied:
Either of commentDirectivesDisabled or commentDirectivesEnabled is fine. Just make sure you switch the flag in the latter case ;)

drpicox replied:
Done!
And flag swap done.

Narretz commented on src/ng/compile.js:
```diff
@@ -1386,6 +1386,61 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+  };
+
+
+  var ignoreCDirectives = false;
```
This one should also be explicit

Narretz commented on src/ng/compile.js:
```diff
@@ -1386,6 +1386,61 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+   * @name $compileProvider#cssClassDirectivesEnabled
+   * @description
+   *
+   * It indicates to the compiler if the current project
```
The description should be something along these lines 
> Whether or not directives on element classes should be compiled. Defaults to `true`.
Calling this function with `false` disables the compilation of directives on element classes for the whole application. This results in a compilation performance gain, as the compiler doesn't have to check element classes when looking for directives. This should however only be used if you are sure that no class directives are used in the application (including any 3rd party directives).

Same change is needed for comment directives, too

drpicox replied:
Good text!
And well seen 3rd party directives.

Narretz commented on test/ng/compileSpec.js:
```diff
@@ -3373,6 +3373,91 @@ describe('$compile', function() {
+      { commentEnabled:false, cssEnabled:false }
+    ], function(config) {
+      describe('commentDirectivesEnabled(' + config.commentEnabled + ')', function() {
+        describe('cssClassDirectivesEnabled(' + config.cssEnabled + ')', function() {
```
Is this second describe needed? Looks to me the description could go into the first describe

drpicox replied:
Ok, I'll change it.

drpicox commented:
Ready!

gkalpak commented on src/ng/compile.js:
```diff
@@ -1386,6 +1386,65 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+   *
+   * @param {boolean} false if the compiler may ignore directives on comments
+   */
+  this.commentDirectivesEnabled = function(value) {
```
We usually make these configuration function getter/setter (and the getters return `self` for chaining).
It makes sense to keep consistent with that.

drpicox replied:
I'll copy the logic from onChangesTtl.

gkalpak commented on src/ng/compile.js:
```diff
@@ -2026,6 +2085,7 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
           }
 
           // use class as directive
+          if (!cssClassDirectivesEnabled) break;
```
I think it is safer to cache the value of the configuration when instantiating the service. Currently it is possible to change this value at runtime, which makes the behavior less predictable.
(Unless we have a complelling usecase for supporting this.)

drpicox replied:
I see, It makes sense. I'll try to update.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -3373,6 +3373,90 @@ describe('$compile', function() {
+      expect(collected).toBe(true);
+    }));
+
+    angular.forEach([
```
You don't need `angular.`.

drpicox replied:
Ops, true! Thanks!

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -3373,6 +3373,90 @@ describe('$compile', function() {
+      { commentEnabled: true, cssEnabled: true },
+      { commentEnabled: true, cssEnabled:false },
+      { commentEnabled:false, cssEnabled: true },
+      { commentEnabled:false, cssEnabled:false }
```
Nitpick: Inconsistent spacing. (Always have a space after `:`, preferrable to have no space after `{` or before `}`.)

petebacondarwin replied:
@mgol will fix this :-)

drpicox replied:
I'm doing it.

gkalpak replied:
@petebacondarwin, just trying to proactively protect @mgol's PR from getting stale :stuck_out_tongue:

drpicox replied:
Ahhh, ok!

On Wed, Jul 27, 2016 at 12:00 PM Georgios Kalpakas <notifications@github.com>
wrote:

> In test/ng/compileSpec.js
> <https://github.com/angular/angular.js/pull/14850#discussion_r72412163>:
>
> > +    it('should collect comment directives by default', inject(function() {
> > +      var html = '<!-- directive: test-collect -->';
> > +      element = $compile('<div>' + html + '</div>')($rootScope);
> > +      expect(collected).toBe(true);
> > +    }));
> > +
> > +    it('should collect css class directives by default', inject(function() {
> > +      element = $compile('<div class="test-collect"></div>')($rootScope);
> > +      expect(collected).toBe(true);
> > +    }));
> > +
> > +    angular.forEach([
> > +      { commentEnabled: true, cssEnabled: true },
> > +      { commentEnabled: true, cssEnabled:false },
> > +      { commentEnabled:false, cssEnabled: true },
> > +      { commentEnabled:false, cssEnabled:false }
>
> @petebacondarwin <https://github.com/petebacondarwin>, just trying to
> proactively protect @mgol <https://github.com/mgol>'s PR from getting
> stale 😛
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/14850/files/f248bd71de405faeca5056e2898d2f924708992d#r72412163>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACoWydUjkPuAVP74kJE6L-cR5geprGoeks5qZywegaJpZM4JB9-c>
> .
>
-- 
Dr. David Ródenas
*Vice-degà *
Col·legi Oficial d’Enginyeria Informàtica de Catalunya

gkalpak commented:
Do we have any (verifiable) numbers on how performance is affected by this?
That would be nice :smile:

drpicox commented:
I did my performance benchmarks using code from my customers (just right click and get source).
I also have used benchpress and numbers looked similar (although I am not sure because I am not familiar with benchpress yet).
I'll try to modify benchpress to enable/disable directives conditionally so we can verify it.

drpicox commented:
I have added a benchpress benchmark. 

- I follow more or less the same structure of the other benchmarks
- I have added query params (they can be set by just clicking links), in order to configure `$compileProvider` in _config_ time.
- I have added a repeater in order to make more benchmark repeats easily (I think that it might be possible to be configured from benchpress, but I didn't found it in a quick glance)
- I have added a chooser for templates: the idea is the compile time varies in function the compiled template. Now I just have added two templates from http://getbootstrap.com/getting-started/#examples , I consider that they can be realistic.

How to run?
- `$ grunt bp_build`
- `$ grunt webserver`
- go to incognito mode
- open [http://localhost:8000/build/benchmarks/bootstrap-compile-bp](http://localhost:8000/build/benchmarks/bootstrap-compile-bp)
- select the kind of directives that  you want to enable
- change if you with the repeats
- select the template
- and click "Loop 25x"
(better if for each test you create a new window)

The numbers that I have obtained are:

- Carousel template:  **~12% of speedup**
  - Compile everything: mean: 206.71ms ± 24% (stddev 49.45)  (min 177.41 / max 332.15)
  - Compile only EA: mean: 183.14ms ± 25% (stddev 45.87) (min 154.76 / max 303.56)

- Theme template: **~10% of speedup**
  - Compile Everything: mean: 615.42ms ± 21% (stddev 126.72) (min 531.03 / max 923.89)
  - Compile only EA: mean: mean: 555.74ms ± 21% (stddev 117.11) (min 474.25 / max 793.57)

gkalpak commented on src/ng/compile.js:
````diff
@@ -1386,6 +1386,75 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+   * $compileProvider.commentDirectivesEnabled(false);
+   * ```
+   *
+   * @param {boolean} false if the compiler may ignore directives on comments
````
You should have the argument name right after the type. Now `false` is interpreted as the argument type and the rest as the description.

gkalpak commented on src/ng/compile.js:
````diff
@@ -1386,6 +1386,75 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+   * ```
+   *
+   * @param {boolean} false if the compiler may ignore directives on comments
+   * @returns {number|object} the current value (or `this` if called as a setter for chaining)
````
`{boolean|object}`

gkalpak commented on src/ng/compile.js:
```diff
@@ -1386,6 +1386,75 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+   * This should however only be used if you are sure that no comment directives are used in the application
+   * (including any 3rd party directives).
+   *
+   * Example:
```
Nitpick: The example is superfluous imo.

drpicox replied:
I liked it (it is just a direct copypaste), but there are no other examples in the file, and `production.ngdoc` has it. So I have deleted it.

gkalpak commented on src/ng/compile.js:
```diff
@@ -1386,6 +1386,75 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
     return TTL;
   };
 
+  var COMMENT_DIRECTIVES_ENABLED = true;
```
Note to self: Using uppercase for non-constant values --> Not intuitive :smile:
(Can be addressed in a separate PR.)

drpicox replied:
I agree, I just copied the TTL example and I'm not sure if there is any naming convention about it. I renamed it to `commentDirectivesEnabledConfig`.

gkalpak commented on docs/content/guide/production.ngdoc:
```diff
@@ -73,3 +73,27 @@ angular.bootstrap(document, ['myApp'], {
+
+## Disable comment and css class directives
+
+By default AngularJS compiles and executes all directives inside comments and element classes. In order to perform this task, angular compiler must look for directives by:
```
Plase limit all lines at 100chars (if possible).

gkalpak commented on docs/content/guide/production.ngdoc:
```diff
@@ -73,3 +73,27 @@ angular.bootstrap(document, ['myApp'], {
+
+Nowadays most of the Angular projects are using only element and attribute directives, and in such projects there is no need to compile comments and classes.
+
+If you are sure that your project only use element directives and attribute directives, and you are not using any 3rd part library that uses directives inside element classes or html comments, you can disable.
```
_use_  -->  _uses_

drpicox committed:
few fixes about flags to compile css class and comment directives conditionally

drpicox commented:
Thanks @gkalpak for comments, I have updated it.

petebacondarwin commented:
LGTM
